regression/test_http_tables.py

```diff
@@ -0,0 +1,156 @@
+        hdr_ref == "*.com" -> chain1;
+        hdr_ref == "http://example.*" -> chain3;
+        hdr_host == "bar*" -> vh5;
+        mark == 1 -> vh5;
+        mark == 2 -> vh4;
+        mark == 3 -> vh3;
+        mark == 4 -> vh2;
+        mark == 5 -> vh1;
+        }
+        """
+    }
+
+    chains = []
+
+    requests_opt = [
+        (('/static/index.html'), ('referer'), ('example.com')),
+        (('/script.php'), ('referer'), ('http://example.com/cgi-bin/show.pl')),
+        (('/foo/example.com'), ('host'), ('test.app.com')),
+        (('/bar/index.html'), ('User-Agent'), ('Mozilla/60.0')),
+        (('/'), ('host'), ('bar.example.com'))
+    ]
+
+    def init_chain(self, (uri, header, value)):
+        ch = chains.base(uri=uri)
+        if header:
+            for request in [ch.request, ch.fwd_request]:
+                request.headers.delete_all(header)
+                request.headers.add(header, value)
+                request.update()
+        self.chains.append(ch)
+
+    def init_server(self, id, srv_response):
+        srv = {}
+        srv['id'] = id
+        srv['type'] = 'deproxy'
+        srv['port'] = str(8000 + id)
+        srv['response'] = 'static'
+        srv['response_content'] = srv_response.msg
+        self.backends.append(srv)
+
+    def init_client(self, id):
+        cl = {}
+        cl['id'] = id
+        cl['type'] = 'deproxy'
+        cl['addr'] = "${tempesta_ip}"
+        cl['port'] = '80'
+        self.clients.append(cl)
+
+    def setUp(self):
+        del(self.chains[:])
+        del(self.backends[:])
+        del(self.clients[:])
+        count = len(self.requests_opt)
+        for i in range(count):
+            self.init_chain(self.requests_opt[i])
+            self.init_server(i, self.chains[i].server_response)
+            self.init_client(i)
+        tester.TempestaTest.setUp(self)
+
+    def start_all(self):
+        self.start_all_servers()
+        self.start_tempesta()
+        self.wait_all_connections()
+        self.start_all_clients()
+
+    def process(self, client, server, chain):
+        client.make_request(chain.request.msg)
+        client.wait_for_response(timeout=5)
+        chain.fwd_request.set_expected()
+        self.assertEqual(server.last_request, chain.fwd_request)
+
+    def test_chains_match(self):
+        """Test for matching rules in HTTP chains: according to
+        test configuration of HTTP tables, requests should arrive to
+        backends in direct order.
+        """
+        self.start_all()
+        count = len(self.chains)
+        for i in range(count):
+            self.process(self.get_client(i),
+                         self.get_server(i),
+                         self.chains[i])
+
+    def test_chains_mark(self):
+        """Test for mark rules in HTTP chains: requests should arrive
+        to backends in reverse order, since mark rules are always processed
+        before match rule (see HTTP tables configuration for current
```

Comment:
Hmm, I missed the point during HTTP chains review in tempesta code. So when we process the next rules:
```
        http_chain {
        hdr_host == "test.app.com" -> chain2;
        hdr_uagent == "Mozilla*" -> chain2;
        hdr_ref == "*.com" -> chain1;
        hdr_ref == "http://example.*" -> chain3;
        hdr_host == "bar*" -> vh5;
        mark == 1 -> vh5;
        mark == 2 -> vh4;
        mark == 3 -> vh3;
        mark == 4 -> vh2;
        mark == 5 -> vh1;
        }
        """
```
with the request 'Host: test.app.com' and mark `2` it will be forwarded to vhost `vh5`.  It can confuse the user, since we have in [docs](https://github.com/tempesta-tech/tempesta/wiki/HTTP-tables):
```
For every HTTP request, HTTP table executes all rule instructions in all linked HTTP chains (beginning from the main chain) sequentially until it finds a match. If no match is found, then the request is dropped.
``` 

Something is not good either documentation or the code. This definitely must be fixed.

Reply:
Yes, we've discussed this in the chat. The problem is that `mark` work at lower layer, when we just get an skb, while HTTP rules work on higher layer, so mark rules must be processed before HTTP rules. We don't want to postpone a job, which we can do now, in performance reasons. Netfilter rules (iptables or nft) work before HttpTables, so it has sense to expect that mark is also processed before HTTP.

I agree with @ikoveshnikov that the point may be confusing for a user, so please document it in the Wiki.

Reply:
Done.